ethereum/contracts/BasicInboundChannelV2.sol

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.5;
+pragma experimental ABIEncoderV2;
+import "./ParachainLightClient.sol";
+import "./BeefyLightClient.sol";
+import "./SimplifiedMMRVerification.sol";
+import "./utils/MerkleProof.sol";
+
+contract BasicInboundChannelV2 {
+    uint256 public constant MAX_GAS_PER_MESSAGE = 100000;
+    uint256 public constant GAS_BUFFER = 60000;
+
+    mapping(bytes32 => uint64) public userNonce;
+
+    BeefyLightClient public beefyLightClient;
+
+    struct Leaf {
+        bytes32 account;
+        uint64 nonce;
+        Message[] messages;
+    }
+
+    struct Message {
+        address target;
+        bytes payload;
+    }
+
+    event MessageDispatched(uint64 nonce, bool result);
+
+    constructor(BeefyLightClient _beefyLightClient) {
+        beefyLightClient = _beefyLightClient;
+    }
+
+    function generateCommitmentHash(Leaf calldata _leaf,bytes32[] calldata _leafProof,bool[] calldata nodeSide)
+    internal pure returns (bytes32)
+    {
+        bytes32 leafNodeHash = keccak256(abi.encode(_leaf));
+
+        return MerkleProof.computeRootFromProofAndSide(leafNodeHash,_leafProof,nodeSide);
+    }
+
+    function submit(
+       Leaf calldata _leaf,
+       bytes32[] calldata leafProof,
+        bool[] calldata nodeSide,
+        ParachainLightClient.ParachainVerifyInput
+            calldata _parachainVerifyInput,
+        ParachainLightClient.BeefyMMRLeafPartial calldata _beefyMMRLeafPartial,
+        SimplifiedMMRProof calldata proof
+    ) public {
+
+        bytes32 commitment = generateCommitmentHash(_leaf,leafProof,nodeSide);
+
+        ParachainLightClient.verifyCommitmentInParachain(
+            commitment,
+            _parachainVerifyInput,
+            _beefyMMRLeafPartial,
+            proof,
+            beefyLightClient
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >= (_leaf.messages.length * MAX_GAS_PER_MESSAGE) + GAS_BUFFER,
+            "insufficient gas for delivery of all messages"
+        );
+
+        processMessages(_leaf);
+    }
+
+    function processMessages(Leaf calldata _leaf) internal { 
+        // User nonce for replay protection
+        require(userNonce[_leaf.account] + 1  == _leaf.nonce , "invalid nonce");
+ 
+        userNonce[_leaf.account] = userNonce[_leaf.account] + 1;
+
+        for (uint256 i = 0; i < _leaf.messages.length; i++) {
+            // Deliver the message to the target
+            (bool success, ) = _leaf.messages[i].target.call{
+                value: 0,
+                gas: MAX_GAS_PER_MESSAGE
+            }(_leaf.messages[i].payload);
+ 
+            emit MessageDispatched(userNonce[_leaf.account], success);
+        }
+    }
+}
```

Comment:
nit: please add a newline at the end of your files.

If you use VS code, the setting `files.insertFinalNewline` will do this automatically for you.